File: lib/test/t_ComposedCopula_std.cxx
//                                               -*- C++ -*-
/**
 *  @brief The test file of class ComposedCopula for standard methods
 *
 *  Copyright 2005-2024 Airbus-EDF-IMACS-ONERA-Phimeca
 *
 *  This library is free software: you can redistribute it and/or modify
 *  it under the terms of the GNU Lesser General Public License as published by
 *  the Free Software Foundation, either version 3 of the License, or
 *  (at your option) any later version.
 *
 *  This library is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU Lesser General Public License for more details.
 *
 *  You should have received a copy of the GNU Lesser General Public License
 *  along with this library.  If not, see <http://www.gnu.org/licenses/>.
 *
 */
#include "openturns/OT.hxx"
#include "openturns/OTtestcode.hxx"

using namespace OT;
using namespace OT::Test;

int main(int, char *[])
{
  TESTPREAMBLE;
  OStream fullprint(std::cout);
  setRandomGenerator();

  try
  {
    // Instantiate one distribution object
    CorrelationMatrix R(3);
    R(0, 1) = 0.5;
    R(0, 2) = 0.25;
    ComposedCopula::DistributionCollection collection(3);
    collection[0] = FrankCopula(3.0);
    collection[1] = NormalCopula(R);
    collection[2] = ClaytonCopula(2.0);
    ComposedCopula copula(collection);
    fullprint << "Copula " << copula << std::endl;
    std::cout << "Copula " << copula << std::endl;

    // Is this copula elliptical ?
    fullprint << "Elliptical distribution= " << (copula.isElliptical() ? "true" : "false") << std::endl;

    // Is this copula continuous ?
    fullprint << "Continuous = " << (copula.isContinuous() ? "true" : "false") << std::endl;

    // Is this copula elliptical ?
    fullprint << "Elliptical = " << (copula.hasEllipticalCopula() ? "true" : "false") << std::endl;

    // Is this copula independent ?
    fullprint << "Independent = " << (copula.hasIndependentCopula() ? "true" : "false") << std::endl;

    // Test for realization of copula
    Point oneRealization = copula.getRealization();
    fullprint << "oneRealization=" << oneRealization << std::endl;

    // Test for sampling
    UnsignedInteger size = 10000;
    Sample oneSample = copula.getSample( size );
    fullprint << "oneSample first=" << oneSample[0] << " last=" << oneSample[size - 1] << std::endl;
    fullprint << "mean=" << oneSample.computeMean() << std::endl;
    UnsignedInteger precision = PlatformInfo::GetNumericalPrecision();
    PlatformInfo::SetNumericalPrecision(4);
    fullprint << "covariance=" << oneSample.computeCovariance() << std::endl;
    PlatformInfo::SetNumericalPrecision(precision);

    // Define a point
    Point point( copula.getDimension(), 0.6 );
    fullprint << "Point= " << point << std::endl;

    // Show PDF and CDF of point
    //Scalar eps(1e-5);
    Point DDF = copula.computeDDF( point );
    fullprint << "ddf     =" << DDF << std::endl;
    Point ddfFD(copula.getDimension());
    fullprint << "ddf (FD)=" << copula.DistributionImplementation::computeDDF(point) << std::endl;
    Scalar PDF = copula.computePDF( point );
    fullprint << "pdf     =" << PDF << std::endl;
    Scalar CDF = copula.computeCDF( point );
    fullprint << "cdf=" << CDF << std::endl;
    Scalar Survival = copula.computeSurvivalFunction(point);
    fullprint << "Survival      =" << Survival << std::endl;
    fullprint << "Survival (ref)=" << copula.computeSurvivalFunction(point) << std::endl;
    Point InverseSurvival = copula.computeInverseSurvivalFunction(0.95);
    fullprint << "Inverse survival=" << InverseSurvival << std::endl;
    fullprint << "Survival(inverse survival)=" << copula.computeSurvivalFunction(InverseSurvival) << std::endl;
    // Get 50% quantile
    Point quantile = copula.computeQuantile( 0.5 );
    fullprint << "Quantile=" << quantile << std::endl;
    fullprint << "CDF(quantile)=" << copula.computeCDF(quantile) << std::endl;

    if (copula.getDimension() <= 2)
    {
      // Confidence regions
      Scalar threshold;
      fullprint << "Minimum volume interval=" << copula.computeMinimumVolumeIntervalWithMarginalProbability(0.95, threshold) << std::endl;
      fullprint << "threshold=" << threshold << std::endl;
      Scalar beta;
      LevelSet levelSet(copula.computeMinimumVolumeLevelSetWithThreshold(0.95, beta));
      fullprint << "Minimum volume level set=" << levelSet << std::endl;
      fullprint << "beta=" << beta << std::endl;
      fullprint << "Bilateral confidence interval=" << copula.computeBilateralConfidenceIntervalWithMarginalProbability(0.95, beta) << std::endl;
      fullprint << "beta=" << beta << std::endl;
      fullprint << "Unilateral confidence interval (lower tail)=" << copula.computeUnilateralConfidenceIntervalWithMarginalProbability(0.95, false, beta) << std::endl;
      fullprint << "beta=" << beta << std::endl;
      fullprint << "Unilateral confidence interval (upper tail)=" << copula.computeUnilateralConfidenceIntervalWithMarginalProbability(0.95, true, beta) << std::endl;
      fullprint << "beta=" << beta << std::endl;
    }

    fullprint << "entropy=" << copula.computeEntropy() << std::endl;
    fullprint << "entropy (MC)=" << -copula.computeLogPDF(copula.getSample(1000000)).computeMean()[0] << std::endl;
    Point mean = copula.getMean();
    fullprint << "mean=" << mean << std::endl;
    ComposedCopula::PointWithDescriptionCollection parameters = copula.getParametersCollection();
    fullprint << "parameters=" << parameters << std::endl;
    // Covariance and correlation
    PlatformInfo::SetNumericalPrecision(4);
    CovarianceMatrix covariance = copula.getCovariance();
    fullprint << "covariance=" << covariance << std::endl;
    CorrelationMatrix correlation = copula.getCorrelation();
    fullprint << "correlation=" << correlation << std::endl;
    CorrelationMatrix spearman = copula.getSpearmanCorrelation();
    fullprint << "spearman=" << spearman << std::endl;
    CorrelationMatrix kendall = copula.getKendallTau();
    fullprint << "kendall=" << kendall << std::endl;
    PlatformInfo::SetNumericalPrecision(precision);
    {
      UnsignedInteger dim = copula.getDimension();
      Scalar x = 0.6;
      Point y(dim - 1, 0.2);
      fullprint << "conditional PDF=" << copula.computeConditionalPDF(x, y) << std::endl;
      fullprint << "conditional CDF=" << copula.computeConditionalCDF(x, y) << std::endl;
      fullprint << "conditional quantile=" << copula.computeConditionalQuantile(x, y) << std::endl;
      Point pt(dim);
      for (UnsignedInteger i = 0; i < dim; ++i) pt[i] = 0.1 * i + 0.05;
      fullprint << "sequential conditional PDF=" << copula.computeSequentialConditionalPDF(pt) << std::endl;
      Point resCDF(copula.computeSequentialConditionalCDF(pt));
      fullprint << "sequential conditional CDF(" << pt << ")=" << resCDF << std::endl;
      fullprint << "sequential conditional quantile(" << resCDF << ")=" << copula.computeSequentialConditionalQuantile(resCDF) << std::endl;
    }

    // Extract a 5-D marginal
    UnsignedInteger dim = 5;
    point = Point(dim, 0.25);
    Indices indices(dim, 0);
    indices[0] = 1;
    indices[1] = 2;
    indices[2] = 3;
    indices[3] = 5;
    indices[4] = 6;
    fullprint << "indices=" << indices << std::endl;
    Distribution margins(copula.getMarginal(indices));
    fullprint << "margins=" << margins << std::endl;
    fullprint << "margins PDF=" << margins.computePDF(point) << std::endl;
    fullprint << "margins CDF=" << margins.computeCDF(point) << std::endl;
    quantile = margins.computeQuantile(0.95);
    fullprint << "margins quantile=" << quantile << std::endl;
    fullprint << "margins CDF(quantile)=" << margins.computeCDF(quantile) << std::endl;
    fullprint << "margins realization=" << margins.getRealization() << std::endl;
    // Tests o the isoprobabilistic transformation
    // General case with normal standard distribution
    fullprint << "isoprobabilistic transformation (general normal)=" << copula.getIsoProbabilisticTransformation() << std::endl;
    // General case with non-normal standard distribution
    collection[0] = SklarCopula(Student(3.0, Point(2, 1.0), Point(2, 3.0), CorrelationMatrix(2)));
    copula = ComposedCopula(collection);
    fullprint << "isoprobabilistic transformation (general non-normal)=" << copula.getIsoProbabilisticTransformation() << std::endl;
    // Special case, independent copula
    collection[0] = SklarCopula(Normal(2));
    collection[1] = IndependentCopula(2);
    collection[2] = NormalCopula(CorrelationMatrix(2));
    copula = ComposedCopula(collection);
    fullprint << "isoprobabilistic transformation (independent)=" << copula.getIsoProbabilisticTransformation() << std::endl;
    {
      dim = copula.getDimension();
      Scalar x = 0.6;
      Point y(dim - 1, 0.2);
      fullprint << "conditional PDF=" << copula.computeConditionalPDF(x, y) << std::endl;
      fullprint << "conditional CDF=" << copula.computeConditionalCDF(x, y) << std::endl;
      fullprint << "conditional quantile=" << copula.computeConditionalQuantile(x, y) << std::endl;
      Point pt(dim);
      for (UnsignedInteger i = 0; i < dim; ++i) pt[i] = 0.1 * i + 0.05;
      fullprint << "sequential conditional PDF=" << copula.computeSequentialConditionalPDF(pt) << std::endl;
      Point resCDF(copula.computeSequentialConditionalCDF(pt));
      fullprint << "sequential conditional CDF(" << pt << ")=" << resCDF << std::endl;
      fullprint << "sequential conditional quantile(" << resCDF << ")=" << copula.computeSequentialConditionalQuantile(resCDF) << std::endl;
    }
    // Special case, single contributor
    collection = Collection<Distribution>(1);
    collection[0] = SklarCopula(Student(3.0, Point(2, 1.0), Point(2, 3.0), CorrelationMatrix(2)));
    copula = ComposedCopula(collection);
    fullprint << "isoprobabilistic transformation (single contributor)=" << copula.getIsoProbabilisticTransformation() << std::endl;
    {
      dim = copula.getDimension();
      Scalar x = 0.6;
      Point y(dim - 1, 0.2);
      fullprint << "conditional PDF=" << copula.computeConditionalPDF(x, y) << std::endl;
      fullprint << "conditional CDF=" << copula.computeConditionalCDF(x, y) << std::endl;
      fullprint << "conditional quantile=" << copula.computeConditionalQuantile(x, y) << std::endl;
      Point pt(dim);
      for (UnsignedInteger i = 0; i < dim; ++i) pt[i] = 0.1 * i + 0.05;
      fullprint << "sequential conditional PDF=" << copula.computeSequentialConditionalPDF(pt) << std::endl;
      Point resCDF(copula.computeSequentialConditionalCDF(pt));
      fullprint << "sequential conditional CDF(" << pt << ")=" << resCDF << std::endl;
      fullprint << "sequential conditional quantile(" << resCDF << ")=" << copula.computeSequentialConditionalQuantile(resCDF) << std::endl;
    }
  }
  catch (TestFailed & ex)
  {
    std::cerr << ex << std::endl;
    return ExitCode::Error;
  }


  return ExitCode::Success;
}

File: python/test/t_ComposedCopula_std.py
#! /usr/bin/env python

from __future__ import print_function
from openturns import *

TESTPREAMBLE()
RandomGenerator.SetSeed(0)

try:
    # Instanciate one distribution object
    R = CorrelationMatrix(3)
    R[0, 1] = 0.5
    R[0, 2] = 0.25
    collection = [FrankCopula(3.0), NormalCopula(R), ClaytonCopula(2.0)]
    copula = ComposedCopula(collection)

    print("Copula ", repr(copula))
    print("Copula ", copula)

    # Is this copula elliptical ?
    print("Elliptical distribution= ", copula.isElliptical())

    # Is this copula continuous ?
    print("Continuous = ", copula.isContinuous())

    # Is this copula elliptical ?
    print("Elliptical = ", copula.hasEllipticalCopula())

    # Is this copula independent ?
    print("Independent = ", copula.hasIndependentCopula())

    # Test for realization of copula
    oneRealization = copula.getRealization()
    print("oneRealization=", repr(oneRealization))

    # Test for sampling
    size = 10000
    oneSample = copula.getSample(size)
    print("oneSample first=", repr(
        oneSample[0]), " last=", repr(oneSample[size - 1]))
    print("mean=", repr(oneSample.computeMean()))
    print("covariance=", repr(oneSample.computeCovariance()))

    # Define a point
    point = Point(copula.getDimension(), 0.6)
    print("Point= ", repr(point))

    # Show PDF and CDF of point
    # Scalar eps(1e-5)
    DDF = copula.computeDDF(point)
    print("ddf     =", repr(DDF))
    PDF = copula.computePDF(point)
    print("pdf     =%.6f" % PDF)
    CDF = copula.computeCDF(point)
    print("cdf=%.6f" % CDF)
    #    Point PDFgr = copula.computePDFGradient( point )
    #    print "pdf gradient     =", PDFgr
    #    Point CDFgr = copula.computeCDFGradient( point )
    #    print "cdf gradient     =", CDFgr
    quantile = copula.computeQuantile(0.95)
    print("quantile=", repr(quantile))
    print("cdf(quantile)=%.6f" % copula.computeCDF(quantile))
    # Get 95% survival function
    inverseSurvival = Point(copula.computeInverseSurvivalFunction(0.95))
    print("InverseSurvival=", repr(inverseSurvival))
    print("Survival(inverseSurvival)=%.6f" %
          copula.computeSurvivalFunction(inverseSurvival))
    print("entropy=%.6f" % copula.computeEntropy())
    # Confidence regions
    if copula.getDimension() <= 2:
        threshold = Point()
        print("Minimum volume interval=",
              copula.computeMinimumVolumeInterval(0.95, threshold))
        print("threshold=", threshold)
        beta = Point()
        levelSet = copula.computeMinimumVolumeLevelSet(0.95, beta)
        print("Minimum volume level set=", levelSet)
        print("beta=", beta)
        print("Bilateral confidence interval=",
              copula.computeBilateralConfidenceInterval(0.95, beta))
        print("beta=", beta)
        print("Unilateral confidence interval (lower tail)=",
              copula.computeUnilateralConfidenceInterval(0.95, False, beta))
        print("beta=", beta)
        print("Unilateral confidence interval (upper tail)=",
              copula.computeUnilateralConfidenceInterval(0.95, True, beta))
        print("beta=", beta)

    mean = copula.getMean()
    print("mean=", repr(mean))
    covariance = copula.getCovariance()
    print("covariance=", repr(covariance))
    parameters = copula.getParametersCollection()
    print("parameters=", repr(parameters))
    # Covariance and correlation
    precision = PlatformInfo.GetNumericalPrecision()
    PlatformInfo.SetNumericalPrecision(4)
    covariance = copula.getCovariance()
    print("covariance=", covariance)
    correlation = copula.getCorrelation()
    print("correlation=", correlation)
    spearman = copula.getSpearmanCorrelation()
    print("spearman=", spearman)
    kendall = copula.getKendallTau()
    print("kendall=", kendall)
    PlatformInfo.SetNumericalPrecision(precision)
    dim = copula.getDimension()
    x = 0.6
    y = [0.2]*(dim-1)
    print("conditional PDF=%.6f" % copula.computeConditionalPDF(x, y))
    print("conditional CDF=%.6f" % copula.computeConditionalCDF(x, y))
    print("conditional quantile=%.6f" %
          copula.computeConditionalQuantile(x, y))
    pt = Point([0.1 * i + 0.05 for i in range(dim)])
    print("sequential conditional PDF=",
          copula.computeSequentialConditionalPDF(point))
    resCDF = copula.computeSequentialConditionalCDF(pt)
    print("sequential conditional CDF(", pt, ")=", resCDF)
    print("sequential conditional quantile(", resCDF, ")=",
          copula.computeSequentialConditionalQuantile(resCDF))

    # Specific to this copula

    # Extract a 5-D marginal
    dim = 5
    point = Point(dim, 0.25)
    indices = Indices(dim, 0)
    indices[0] = 1
    indices[1] = 2
    indices[2] = 3
    indices[3] = 5
    indices[4] = 6
    print("indices=", repr(indices))
    margins = copula.getMarginal(indices)
    print("margins=", repr(margins))
    print("margins PDF=%.6f" % margins.computePDF(point))
    print("margins CDF=%.6f" % margins.computeCDF(point))
    quantile = margins.computeQuantile(0.95)
    print("margins quantile=", repr(quantile))
    print("margins CDF(quantile)=%.6f" % margins.computeCDF(quantile))
    print("margins realization=", repr(margins.getRealization()))
    # Tests the isoprobabilistic transformation
    # General case with normal standard distribution
    print("isoprobabilistic transformation (general normal)=",
          copula.getIsoProbabilisticTransformation())
    # General case with non-normal standard distribution
    collection[0] = SklarCopula(Student(
        3.0, Point(2, 1.0), Point(2, 3.0), CorrelationMatrix(2)))
    copula = ComposedCopula(collection)
    print("isoprobabilistic transformation (general non-normal)=",
          copula.getIsoProbabilisticTransformation())
    dim = copula.getDimension()
    x = 0.6
    y = [0.2]*(dim-1)
    print("conditional PDF=%.6f" % copula.computeConditionalPDF(x, y))
    print("conditional CDF=%.6f" % copula.computeConditionalCDF(x, y))
    print("conditional quantile=%.6f" %
          copula.computeConditionalQuantile(x, y))
    pt = Point([0.1 * i + 0.05 for i in range(dim)])
    print("sequential conditional PDF=",
          copula.computeSequentialConditionalPDF(pt))
    resCDF = copula.computeSequentialConditionalCDF(pt)
    print("sequential conditional CDF(", pt, ")=", resCDF)
    print("sequential conditional quantile(", resCDF, ")=",
          copula.computeSequentialConditionalQuantile(resCDF))
    # Special case, independent copula
    collection[0] = SklarCopula(Normal(2))
    collection[1] = IndependentCopula(2)
    collection[2] = NormalCopula(CorrelationMatrix(2))
    copula = ComposedCopula(collection)
    print("isoprobabilistic transformation (independent)=",
          copula.getIsoProbabilisticTransformation())
    dim = copula.getDimension()
    x = 0.6
    y = [0.2]*(dim-1)
    print("conditional PDF=%.6f" % copula.computeConditionalPDF(x, y))
    print("conditional CDF=%.6f" % copula.computeConditionalCDF(x, y))
    print("conditional quantile=%.6f" %
          copula.computeConditionalQuantile(x, y))
    pt = Point([0.1 * i + 0.05 for i in range(dim)])
    print("sequential conditional PDF=",
          copula.computeSequentialConditionalPDF(pt))
    resCDF = copula.computeSequentialConditionalCDF(pt)
    print("sequential conditional CDF(", pt, ")=", resCDF)
    print("sequential conditional quantile(", resCDF, ")=",
          copula.computeSequentialConditionalQuantile(resCDF))
    # Special case, single contributor
    collection = [SklarCopula(Student(
        3.0, Point(2, 1.0), Point(2, 3.0), CorrelationMatrix(2)))]
    copula = ComposedCopula(collection)
    print("isoprobabilistic transformation (single contributor)=",
          copula.getIsoProbabilisticTransformation())
    dim = copula.getDimension()
    x = 0.6
    y = [0.2]*(dim-1)
    print("conditional PDF=%.6f" % copula.computeConditionalPDF(x, y))
    print("conditional CDF=%.6f" % copula.computeConditionalCDF(x, y))
    print("conditional quantile=%.6f" %
          copula.computeConditionalQuantile(x, y))
    pt = Point([0.1 * i + 0.05 for i in range(dim)])
    print("sequential conditional PDF=",
          copula.computeSequentialConditionalPDF(pt))
    resCDF = copula.computeSequentialConditionalCDF(pt)
    print("sequential conditional CDF(", pt, ")=", resCDF)
    print("sequential conditional quantile(", resCDF, ")=",
          copula.computeSequentialConditionalQuantile(resCDF))

    # test ComposedCopula.getMarginal in reverse
    copula = ComposedCopula([IndependentCopula(2), NormalCopula(2)])
    print(copula.getMarginal([3, 2, 1, 0]))

except:
    import sys
    print("t_ComposedCopula.py", sys.exc_info()[0], sys.exc_info()[1])
